source/common/http/conn_pool_grid.cc

```diff
@@ -316,21 +330,33 @@ void ConnectivityGrid::markHttp3Broken() { http3_status_tracker_.markHttp3Broken
 
 void ConnectivityGrid::markHttp3Confirmed() { http3_status_tracker_.markHttp3Confirmed(); }
 
-void ConnectivityGrid::onDrainReceived() {
+bool ConnectivityGrid::isIdle() const {
+  // This is O(n) but the function is constant and there are no plans for n > 8.
+  bool idle = true;
+  for (const auto& pool : pools_) {
+    idle &= pool->isIdle();
+  }
+  return idle;
+}
+
+void ConnectivityGrid::onIdleReceived() {
   // Don't do any work under the stack of ~ConnectivityGrid()
   if (destroying_) {
     return;
   }
 
-  // If not all the pools have drained, keep waiting.
-  ASSERT(drains_needed_ != 0);
-  if (--drains_needed_ != 0) {
-    return;
+  bool drained = false;
+  if (drains_needed_ > 0) {
+    if (--drains_needed_ == 0) {
```

Comment:
draining is a terminal state (assuming no API use violations). Idle can happen repeatedly. 

So with how I've written this grid code, the grid doesn't report itself as drained or idle unless all of the pools it owns report as such.

For draining, these will happen one by one, and a pool can't un-drain.

With idle, if a sub-pool was idle, but then gets a request, it will be not-idle anymore, so I need to check all of them everytime any one of them becomes idle to figure out the current state.

Reply:
Ok, I think I understand this and I think it's correct. (No surprise there :>) But let me make sure that I understand. I was potentially worried about this code:
```
  if (drains_needed_ > 0) {
    if (--drains_needed_ == 0) {
      // All the pools have drained. Notify drain subscribers.
      drained = true;
    }
  }

  if (drained || isIdle()) {
    for (auto& callback : idle_callbacks_) {
      callback();
    }
  }
}```
  
in which `onIdleReceived()` decremented `drains_needed_`. I was concerned that the same pool might call `onIdleReceived()` twice and then we would invoke the callbacks too early.

But I think that's not possible because once `startDraining()` is called it's not possible for one of the underlying pools to transition from idle to not-idle, and hence it's not possible to invoke `onIdleReceived()` twice. Do I have that right? If so, this looks good to me!

(I do wonder if it wouldn't be easier to only call `isIdle()` down on 356 and ditch `drains_needed_` completely. O(2) is pretty small so I'm not sure how much we safe by tracking idle explicitly. But this is a minor point.)

Reply:
Yes, your understanding is correct.

Hmm, yeah, I think you're right that we don't need to do both checks. I've adapted this code several times through different iterations of this PR, and I think that this isn't needed anymore. I'll clean it up.

Reply:
Awesome! This looks great. Thanks for walking me through the logic! (and thanks for doing this)